adds the feature to animate a color property on the a-animation compo…

JohnRodney opened the pull request:
**Description:**
  Adds the support for using `a-animation` components in order to animate colors. #733
**Changes proposed:**
-  Adds the detection of a color as an attribute
-  Adds a partialSetAttribute for color
-  Adds a conversion from rgb object to hex (maybe these belong in utils?)
-  Adds two unit tests for start and finish to ensure the color is properly tweened

…nent

ngokevin commented on src/utils/coordinates.js:
```diff
@@ -48,7 +48,7 @@ module.exports.stringify = stringify;
  */
 module.exports.isCoordinate = function (value) {
-  return regex.test(value);
+  return value[0] !== '#' && regex.test(value);
```
Color hexes were passing the regex test?

JohnRodney replied:
Yes

ngokevin replied:
So coordinates will always have whitespaces in between the numbers. Colors won't. I see that the whitespace characters uses star (`*`) rather than (`+`) which will match even 0 whitespace characters. Need to change those to `+` to force at least 1 whitespace character.

try `/\s*(-?\d*\.{0,1}\d+)\s+(-?\d*\.{0,1}\d+)\s+(-?\d*\.{0,1}\d+)\s*/;`

JohnRodney replied:
This has been updated as well thanks for the help on the regex.

ngokevin commented on tests/core/a-animation.test.js:
```diff
@@ -118,6 +118,34 @@ suite('a-animation', function () {
     });
   });
 
+  suite('component color animation', function () {
```
Would be nice to have tests for:

- Shorthand `#fff`
- Capital `#FFF`
- Interpolated value. Check that between the animation, the color is somewhere between the start and end.

JohnRodney replied:
These test cases have been added.

ngokevin replied:
Thanks!

dmarcos commented on src/core/a-animation.js:
```diff
@@ -463,3 +479,35 @@ function strToBool (str) {
+ * @returns {string}
+ */
+function componentToHex (color) {
+  var hex = color.toString(16);
```
Should we check if color < 0 or > 255 and clamp it to 0-255?

JohnRodney replied:
@dmarcos @ngokevin Hey this could be clamped in the function that calls this by doing a:
`  return Math.floor(Match.min(Math.abs(color), 255) * 255);`
inside of the convertToIntegerColor() function.  Would this be acceptable?  Or would you prefer a separate clamp function?

JohnRodney replied:
This is also possibly unneeded as the value that is passed comes from a THREE.Color which is already tested to be within 0 and 1 from a hex value correct?

ngokevin commented on tests/core/a-animation.test.js:
```diff
@@ -118,6 +118,78 @@ suite('a-animation', function () {
+      var color;
+      this.animationEl.tween.update(this.startTime + 500);
+      color = this.el.getComputedAttribute('color');
+      assert.isAbove(color, '#000000');
```
Very nice that you can assert like this.

ngokevin commented on src/core/a-animation.js:
```diff
@@ -5,6 +5,7 @@ var parseProperty = require('./schema').parseProperty;
 var registerElement = require('./a-register-element').registerElement;
 var TWEEN = require('tween.js');
 var utils = require('../utils/');
+var THREE = require('three');
```
do ```../lib/three```

dmarcos commented on src/core/a-animation.js:
```diff
@@ -363,6 +364,8 @@ function getAnimationValues (el, attribute, dataFrom, dataTo, currentValue) {
     getForCoordinateComponent();
   } else if (['true', 'false'].indexOf(dataTo) !== -1) {
     getForBoolean();
+  } else if (dataTo && dataTo[0] === '#') {
```
We should also use the schema to infer the color type and not just based on the '#' prefix. We could do:
````html
<a-animation attribute="material.color" from="blue" to="red" dur="4000" easing="ease-out"></a-animation>
````

JohnRodney replied:
Good idea.  Can you point me to the right direction on this?  Is there already a currently existing way to check if these are valid colors or would this be something I need to add?

ngokevin replied:
I think ```#``` is more flexible. Using the schema would not work in cases like:

```html
<a-box color="#CCC">
  <a-animation attribute="color" to="#FFF"></a-animation>
</a-box>
```

dmarcos replied:
The THREE.Color class does it for you.
http://threejs.org/docs/#Reference/Math/Color
You can do ```new THREE.Color('red').getHexString()````

dmarcos replied:
API consistency is important. Let's do it now.

JohnRodney replied:
Alright I'm about to do my hour drive home and then I'll try to parse through the schema code to make these adjustments.  May not get it done until tomorrow.  Thanks for all the feedback. @dmarcos @ngokevin

dmarcos replied:
@JohnRodney Thanks for the effort. This is good stuff.

JohnRodney replied:
@dmarcos @ngokevin.  Alright on the drive home I was thinking the catching else statement does an automatic type conversion to number which will turn out to be a NaN property on the object.  So instead of checking prop or color I figured we can just check for `isNaN(dataTo)` and then feed it directly to the Three.js conversions.  This will always get a color (white) if an invalid color is passed and will still throw a warning from three.js when the color is not valid.

This gets around checking specific color formats and allows Three.js to handle improper input. Is there any reason you would think this would fail?

JohnRodney commented:
I think all feedback is resolved and I went through a refactor phase.  It should be ready to squash the commits if everything else is approved.

ngokevin commented:
@dmarcos This looks pretty good. He made it not explicitly check for color, just let it fall through the if/else statements, saving about probably 50 LOC. I think even if we add more interpolation types, this should still hold.

dmarcos commented on src/core/a-animation.js:
```diff
@@ -363,6 +364,8 @@ function getAnimationValues (el, attribute, dataFrom, dataTo, currentValue) {
     getForCoordinateComponent();
   } else if (['true', 'false'].indexOf(dataTo) !== -1) {
     getForBoolean();
+  } else if (isNaN(dataTo)) {
```
Does this work for the following case? Can we add it to the tests?
````html
<a-animation attribute="material.color" from="red" to="blue"></a-animation>
````

JohnRodney replied:
Good catch. I'll look at it in the morning.

ngokevin replied:
It works.

ngokevin commented on src/core/a-animation.js:
```diff
@@ -391,7 +398,14 @@ function getAnimationValues (el, attribute, dataFrom, dataTo, currentValue) {
     }
     from[attribute] = parseProperty(from[attribute], schema[componentPropName]);
     to[attribute] = parseProperty(dataTo, schema[componentPropName]);
```
If our `color` property type parser converted to `THREE.Color` I think everything would work out of the box. But I'm not sure if that would break things.

ngokevin commented on src/core/a-animation.js:
```diff
@@ -391,7 +398,14 @@ function getAnimationValues (el, attribute, dataFrom, dataTo, currentValue) {
+      to[attribute] = new THREE.Color(to[attribute]);
+    }
     partialSetAttribute = function (value) {
+      if (attribute.indexOf('color') !== -1) {
```
You could do ```schema[componentPropName].type === 'color'```, but I'm still looking at the whole thing.

ngokevin commented on src/core/a-animation.js:
```diff
@@ -138,7 +139,11 @@ module.exports.AAnimation = registerElement('a-animation', {
 
         // Create Tween.
-        return new TWEEN.Tween(cloneValue(from))
+        if (from['material.color']) {
```
How come we need this?

JohnRodney replied:
I'm not sure.  I am sure it was added for a reason because of some case to lose an object reference since all object assignments are references in JS.  I haven't dug in to see if this is a shallow clone or a deep clone.  

I mainly made this commit to share how I was able to 'fix' the issue, but I wasn't completely sure why these changes fixed it. I know its just an object reference somewhere.

JohnRodney commented:
@ngokevin @dmarcos Updated with a new approach to support both `color` and `material.color`.  I feel better about this approach.  Will add in some unit tests to ensure no regressions are introduced around the 'material.color' property.

ngokevin commented on src/core/a-animation.js:
```diff
@@ -356,13 +357,14 @@ function getAnimationValues (el, attribute, dataFrom, dataTo, currentValue) {
   var to = {};
-
-  if (attributeSplit.length === 2) {
+  if (attributeSplit.length === 2 && attributeSplit.indexOf('color') === -1) {
```
Note that `material.color` is only a single test case. Properties can be named anything and still be of type color. I'll try to help find a solution.

JohnRodney replied:
are you saying something with a color property could be passed that did not contain the word color in it?

ngokevin replied:
Yup. Anyone can create components with whatever schema they want. So I could do:

```
schema: {
  body: {type: 'color'}
}
```

JohnRodney replied:
Ok so exact matching the scheme type should be safe here and then the rest should still work as far as the tweets are concerned.  I'm using the passed attributed to set when there are more than one in attributeSplit

ngokevin replied:
Ah, that's interesting. Seems like that might work. It's not ideal we need to special-case colors, but I'm feel okay with this.

JohnRodney replied:
Elaborate on special case colors?  I don't mind getting the solution all the way done rather than leaving something for someone else.

ngokevin replied:
As in we have to a special check for colors. Though I don't have a better simple solution right now.

dmarcos replied:
We need to check the schema. If we have for example a phong shader you will have three properties: diffuse, ambient, specular that are all colors.

JohnRodney replied:
Ok I have something like: 

```
function isColor() {
  if (el.mappings && attributeSplit.length === 1 && !el.components[attribute]){
    return el.components[el.mappings[attribute].split('.')[0]].schema[attribute].type === 'color';
  } else if(attributeSplit.length > 1 && el.components[attributeSplit[0]]) {
    return el.components[attributeSplit[0]].schema[attributeSplit[1]].type === 'color';
  }
  return false;
}
```

There has to be an easier way to access the schema?
The reason I had to do this was to access the correct schema on a single attribute 'color' I had to access the mappings because there was no color schema it was under 'material.color' as that is the default map.

ngokevin replied:
If the attribute can be split, then you can use `el.components[attrSplit[0]].schema[attrSplit[1]].type === 'color'`.

Else you can let it fall thru the if/else statements with the `isNaN`.

JohnRodney replied:
@ngokevin In this case `attribute = 'color'` so `attrSplit[1]  = undefined`.
The `mapping[attribute] = 'material.color'` so `mapping[attribute].split('.')[0]] = 'material'` aka the component name.

ngokevin replied:
If the attribute is `color`, then the first clause will fail because there's nothing to split on. Thus you just move on to the `isNaN(dataTo)` and do color interpolation there.

If the attribute is `material.color`, then you know you can check the schema, right?

JohnRodney replied:
You are right that will support both.  And make all the test cases for color animation currently commited still work.  

Then we would just have to figure out how to add `el.components.material.schema.color.type = 'color'` to the generated entity to test the cases of `material.color` or `material.specular`

ngokevin replied:
You don't have to add anything. Just set the material component onto the entity.

```el.setAttribute('material', 'shader: flat; color: #FFF');```

JohnRodney replied:
Now that I look at it I had tried that method earlier but it breaks 4 tests because I'm guessing in certain test cases the mapping doesn't exist so its not safe to check for it as the tests will try to call a property on an undefined object.

ngokevin replied:
I'd ignore everything about mappings. That's part of what is currently not core code (e.g., `<a-box>`). Though it should be able to interpolate attributes that it detects are colors (e.g., `color="#FFFFFF"`, but that doesn't require looking up schema.

JohnRodney replied:
Alright took your suggestion and pushed another version up with the fallback to isNaN and a isColor function that checks to make sure schemas are defined so that older tests don't fail with this introduction.

dmarcos commented on src/core/a-animation.js:
```diff
@@ -373,6 +375,17 @@ function getAnimationValues (el, attribute, dataFrom, dataTo, currentValue) {
+   * @return {bool} if the schema is of type color
+   */
+  function isColor () {
+    if (el.components[attributeSplit[0]] && el.components[attributeSplit[0]].schema) {
```
You can clean this up a little bit:
````javascript
function isColor () {
  var componentName = attributeSplit[0];
  var propertyName = attributeSplit[1];
  var component = el.components[componentName];
  var schema = component && component.schema;
  if (schema) { return schema[propertyName].type === 'color' }
  return true;
}

dmarcos commented on src/core/a-animation.js:
```diff
@@ -373,6 +375,17 @@ function getAnimationValues (el, attribute, dataFrom, dataTo, currentValue) {
   };
 
   /**
+   * Match the sceme type to color
```
typo sceme -> schema

dmarcos commented on src/core/a-animation.js:
```diff
@@ -356,13 +357,14 @@ function getAnimationValues (el, attribute, dataFrom, dataTo, currentValue) {
   var to = {};
-
-  if (attributeSplit.length === 2) {
+  if (attributeSplit.length === 2 && !isColor()) {
```
It's better to be explicit that let it pass through:
````javascript
if (attributeSplit.length === 2) {
  if (isColor()) { getForColorComponent(); }
  else {  getForComponentAttribute() }
}
````

dmarcos commented on tests/core/a-animation.test.js:
```diff
@@ -40,6 +40,51 @@ function setupAnimation (animationAttrs, cb, elAttrs) {
+ *  @param {from} expects a color format of white
+ *  @param {to} expects a color format of black
+ */
+function generateColorAnimationTest (description, from, to) {
```
We need a test for the case:
````html
<a-animation attribute="material.color" from="red" to="blue"></a-animation>
````

JohnRodney commented:
Updated with refactors as suggested and added the red to blue test case.  Should probably go through another refactor phase and then add an example and update the docs?

Will probably update a bit more tomorrow.

dmarcos commented on src/core/a-animation.js:
```diff
@@ -373,6 +379,19 @@ function getAnimationValues (el, attribute, dataFrom, dataTo, currentValue) {
+    var propertyName = attributeSplit[1];
+    var component = el.components[componentName];
+    var schema = component && component.schema;
+    if (schema) { return schema[propertyName].type === 'color'; }
```
we can simplify the return statment:
````javascript
return schema && schema[propertyName].type === 'color';
````

JohnRodney replied:
This throws an error when schema[propertyName] is undefined so it would have to be:

```
  return schema && schema[propertyName] && schema[propertyName].type === 'color';
```

Would protect from all. Not sure how often the schema is undefined I think its only when given bad input, but probably better than letting an error get thrown?

dmarcos replied:
Now that I think about this. We could animate a component that it's not defined on the element yet (before the animation runs like)
````html
<a-entity geometry="..."><a-animation attribute="material.color" from="red" to="blue"></a-animation></a-entity>
````
Is this case covered? I think you would not be able to check ```el.components``` to retrieve the schema.

ngokevin replied:
Worth testing, though I suspect it will set the `from` before starting?

JohnRodney replied:
As far as I have seen yes it sets the from as the default on the image even if a color hasn't been set.

JohnRodney replied:
@ngokevin Did you ever get a chance to check out the branch?  Also was wondering if going in this direction for the animation should maybe be rethought?  Having to add special cases for each animation type seems unintuitive.  Couldn't we do a more effective remapping of primitives to their object by passing a reference?

Just worried this branching if of detecting type is going to get uglier as we try to support different interpolation types.  I built out something similar in a repo where I was trying to essentially build out an aframe-like framework (before I heard of aframe) in meteor.  

Through this I had build out a transitioner object that would take a payload and tween any object with numeric values.  As long as you passes a string describing it such as `mesh.material.specular` and a ending object `{ r: 1, g: 0, b: .4 }`.

Here is a reference to the code and how I approached it.  Not sure if it would help https://github.com/poetic/poetic-vr-boilerplate/blob/master/utils-animate.js.

dmarcos replied:
@JohnRodney Yes, I think this PR is very ad-hoc for the color type. The animation code is one of the most confusing areas in aframe and any ideas to improve it are more than welcome. In your approach, Would you pass the variable you want interpolated and the end value? Do you infer the type from the end value?

ngokevin replied:
- tween.js can already tween any object with numeric values.
- Might want to refactor if it gets any more complicated than this, will re-evaluate. Are there any other special cases we need to consider? Colors might be the last big special case.

Code looks OK for this scope. Minimal changes, tests pass, things work :white_check_mark:

JohnRodney replied:
@dmarcos Yea I approached it from the idea that it would tween values from your passed object and whatever it's value already was.  As for type checking it was a pure number tween so it would just diff the values calculate delta from last from and decide the amount to tween based on the current framerate.

@ngokevin Understood on tween.js.  I ended up writing that because it didn't take much code to accomplish tween js w/o the overhead of another library.  

I am still not positive this would even work on 'material.specular' as we discussed earlier.  I need to dig in a bit more to how to setup a color value for specular and then try to see if we can animate that as well.

ngokevin commented on src/core/a-animation.js:
```diff
@@ -373,6 +379,19 @@ function getAnimationValues (el, attribute, dataFrom, dataTo, currentValue) {
   };
 
   /**
+   * Match the scema type to color
```
typo schema

JohnRodney replied:
lol I've mistyped that one twice now.  Updating the commit.

ngokevin commented on src/core/a-animation.js:
```diff
@@ -463,3 +498,34 @@ function strToBool (str) {
+
+/**
+ * Converts a rgb object into a hex string
+ * @param {color} object { r: 1, g: 1, b: 1 }
```
reversed jsdoc `@param {object} color - {r: 1, g: 1, b: 1}`

JohnRodney replied:
resolved

ngokevin commented:
Tested, works on both ECS box and primitive-style box:

![box](https://cloud.githubusercontent.com/assets/674727/14259829/9a793a72-fa5d-11e5-8bf3-da28b491b93e.gif)

![cube](https://cloud.githubusercontent.com/assets/674727/14259832/9b578b42-fa5d-11e5-8a04-ea045a6736c8.gif)

(In that, I discovered an issue that is not related to this PR: https://github.com/aframevr/aframe/issues/1334)

ngokevin commented:
I think we're ready to squash, and @dmarcos to sign off.

JohnRodney commented:
Excellent did we want to update the docs in animation.md to explain color animation support with a short snippet?

ngokevin commented:
Yeah that'd be great. An `test-animations-color` example would be cool too.

JohnRodney commented:
I am guessing a `test-animations-color` is different from an animation example?  It looks like `test` example namespace folders are all ones that provide stats and seem to be attempting to put some burden on the system?

ngokevin commented:
We need to organize/rename the examples. But `test-` examples, to me, are examples that you check to make sure a particular feature works. They all have stats enabled on the scene at the moment so we can check mobile stats, but we could re-eval that.

JohnRodney commented:
Sounds good I'll follow the examples of the other test-examples and put some color on it.  I had been playing around with some pulse animations that domino event emissions. Just to see if too many animations on color would cause a problem.  Never really saw an issue, but I didn't add stats to the scene. 

![out](https://cloud.githubusercontent.com/assets/11075982/14265910/f15a7f64-fa89-11e5-9645-b33898705637.gif)

cvan commented:
woo! this is awesome! :tada:

JohnRodney commented:
@ngokevin Sorry about the delay I got really sick last week.  I added the readme entry and a test-animate-color.  Once this is ready I can add the test we talked about to the other PR that ensures a from color is correctly retrieved when a animating a color and the from is not defined in the a-animation.

ngokevin commented on examples/test-animate-color/index.html:
```diff
@@ -0,0 +1,25 @@
+  <body>
+    <a-scene stats="true">
+      <a-sky color="#afafaf"></a-sky>
+      <a-box color="white" position="2 1 -10">
```
Can you make a few of these `<a-entity>` syntax?

ngokevin commented on tests/core/a-animation.test.js:
```diff
@@ -550,4 +606,55 @@ suite('a-animation', function () {
+    var attribute = 'material.color';
+    setup(function (done) {
+      var self = this;
+      var elAttrs;
```
Can you build this object in just one line?

```js
var elAttrs = { 
  material: {}
};
```

ngokevin commented:
Thanks, hope you're better!

dmarcos commented:
@JohnRodney Is there any work left here?

JohnRodney commented:
@dmarcos yea I just need to alter @ngokevin suggestions.  I'm digging into now.  Not sure if you guys seen the news but weather has been pretty crazy in Houston the last few days.  I'll be pushing an update up shortly.

dmarcos commented:
@JohnRodney Thanks! I hope the sun shines soon over there.

JohnRodney commented:
@dmarcos We all do here too haha.

@ngokevin I adjusted the requested changes.  The only thing that may be missing I can think of now is a link to the example on the index page that is on localhost:9000 when you start the server.  I'm guessing I should probably add this yes?

ngokevin commented on docs/core/animations.md:
````diff
@@ -73,6 +73,16 @@ We can animate numeric attributes as well. For example, we can animate the inten
+
+```html
+<a-entity id="blushing-cube" geometry="primitive: box">
+  <a-animation attribute="color" begin="white" to="red" dur="1000"></a-animation>
````
Should be `attribute="material.color"`

`begin="white"` should be `from="white"`

ngokevin commented on docs/core/animations.md:
```diff
@@ -73,6 +73,16 @@ We can animate numeric attributes as well. For example, we can animate the inten
 
+### Color Attributes
+
+We can animate color attributes as well. For example, we can animate a box from white to red.
```
I'd add `We can animate any component property that has a `color` type as well.`

ngokevin commented on examples/test-animate-color/index.html:
```diff
@@ -0,0 +1,28 @@
+  <body>
+    <a-scene stats="true">
+      <a-sky color="#afafaf"></a-sky>
+      <a-box color="white" position="2 1 -10">
```
Is it possible to use the example you posted?

![](https://cloud.githubusercontent.com/assets/11075982/14265910/f15a7f64-fa89-11e5-9645-b33898705637.gif)

ngokevin replied:
This one is really cool

JohnRodney replied:
@ngokevin I used setTimeouts in that to stagger the animation emission of events.  If you guys are ok with that.  Some repo's are real against timeouts so I wasn't sure if I should use it as an example.

JohnRodney replied:
If that is cool I will recreate that example again.

ngokevin replied:
You can use the animation 'begin' attribute to delay animation starts

JohnRodney replied:
Alright I'll look into that and get the working example to run around that method.  Getting caught up at work today too so probably look for this adjustment tomorrow.

ngokevin commented:
OK, that should be the last of it!

JohnRodney commented:
@ngokevin alright I updated the demo to a little more fancy of a demo.  I also opened an issue because you can't use begin delays to stagger because they calculate the begin delay on a repeat meaning the "stagger" animation eventually gets out of sync.  

This can be updated to remove the intervals once the bug is fixed there.

ngokevin commented:
Awesome! I'll merge this once I get to the office.

JohnRodney commented:
Cool once you give it one more look over and are ready let me know and I'll squash the commits

ngokevin commented on examples/test-animate-color/index.html:
```diff
@@ -0,0 +1,153 @@
+        <a-animation attribute="color" dur="10000" from="#afafaf" to="#000000" repeat="indefinite">
+      </a-sky>
+
+      <a-entity position="4.5 1 -3" class="pulse" geometry="primitive: box; depth: 0.5; height: 1; width: 0.5" material="color: #5698Af">
```
You can use mixins to clean up.

```
html
<a-assets>
  <a-mixin id="box" geometry="primitive: box..." material="..."></a-mixin>
  <a-mixin id="pulse" attribute="material.color" from="#fff" to="#5698Af" dur="1000" begin="pulse"></a-mixin>
</a-assets>

<a-entity mixin="box"><a-animation mixin="pulse"></a-animation></a-entity>
<a-entity mixin="box"><a-animation mixin="pulse"></a-animation></a-entity>
```

ngokevin commented on examples/test-animate-color/index.html:
```diff
@@ -0,0 +1,153 @@
+<!DOCTYPE html>
```
We organized the examples, so you have to move this to ```examples/test/animate-color```

And change to ```../../../dist/aframe.js```

ngokevin commented on examples/test-animate-color/index.html:
```diff
@@ -0,0 +1,153 @@
+    <script src="../../dist/aframe.js"></script>
+  </head>
+  <body>
+    <a-scene stats="true">
```
Can just do `<a-scene stats>`

ngokevin commented on examples/test-animate-color/index.html:
```diff
@@ -0,0 +1,153 @@
+  </body>
+
+  <script>
+    window.onload = function() {
```
Our linter isn't running on examples yet, but:

- Space after `function` (`function ()`)
- Space around operators (`length / 2`)

Might be better to wait for `scene` `loaded` event too.

ngokevin commented on examples/test-animate-color/index.html:
```diff
@@ -0,0 +1,153 @@
+  </head>
+  <body>
+    <a-scene stats="true">
+      <a-sky color="#000000">
```
It is `#000` here, but the `from` is different.

JohnRodney commented:
@ngokevin adjusted to mixins, new file structure and stats attribute.

ngokevin commented on examples/test/animate-color/index.html:
```diff
@@ -0,0 +1,79 @@
+        <a-animation attribute="color" dur="10000" from="#afafaf" to="#000000" repeat="indefinite">
+      </a-sky>
+
+      <a-entity position="4.5 1 -3" class="pulse" mixin="box"><a-animation mixin="pulse"></a-animation></a-entity>
```
You don't need the class anymore because your `selectorAll` can be `[mixin="box"]`

JohnRodney replied:
done

ngokevin commented:
Thanks so much!!! That was a long process, but everything works, and your example is pretty cool and running at 60fps  :)
